gtfu/src/main/java/gtfu/tools/StaticGTFSToBucket.java

```diff
@@ -0,0 +1,154 @@
+package gtfu.tools;
+
+import gtfu.ProgressObserver;
+import gtfu.ConsoleProgressObserver;
+import gtfu.Util;
+import gtfu.Debug;
+import gtfu.EmailFailureReporter;
+import gtfu.FailureReporter;
+import gtfu.Recipients;
+
+import java.util.zip.ZipFile;
+import java.util.zip.ZipEntry;
+import java.nio.charset.StandardCharsets;
+import java.io.IOException;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.util.Enumeration;
+
+/**
+ * Uploads agency static GTFS files to GCloud Bucket
+ */
+public class StaticGTFSToBucket {
+    /**
+     * Uploads static GTFS for a single agency to the specified bucket
+     * @param agencyID   Agency
+     * @param bucketName GCloud bucket name
+     * @param cacheDir   Cache location for storing zip files
+     */
+    public static void StaticGTFSToBucket(String agencyID, String bucketName, String cacheDir) {
+        String[] agencyIDList = {agencyID};
+        StaticGTFSToBucket(agencyIDList, bucketName, cacheDir);
+    }
+
+    /**
+     * Uploads static GTFS for each agency to the specified bucket
+     * @param agencyIDList A list of agencyIDs
+     * @param bucketName   GCloud bucket name
+     * @param cacheDir     Cache location for storing zip files
+     */
+    public static void StaticGTFSToBucket(String[] agencyIDList, String bucketName, String cacheDir) {
+        AgencyYML yml = new AgencyYML();
+
+        Recipients r = new Recipients();
+        GCloudStorage gcs = new GCloudStorage();
+        String[] recipients = r.get("error_report");
+
+        FailureReporter reporter = new EmailFailureReporter(recipients, "Update Trip Names Report");
+        Util.setReporter(reporter);
+
+        File file = new File(cacheDir);
+
+        if (!file.exists()) {
+            file.mkdir();
+        }
+
+        reporter.addLine("Uploaded GTFS data for:");
+
+        for (String agencyID : agencyIDList) {
+            reporter.addLine("- " + agencyID);
+            String name = cacheDir + "/" + agencyID;
+            file = new File(name);
+
+            if (!file.exists()) {
+                file.mkdir();
+            }
+
+            String gtfsURL = yml.getURL(agencyID);
+
+            try {
+                String cl = Util.getResponseHeader(gtfsURL, "Content-Length");
+
+                if (cl == null) {
+                    cl = "0";
+                }
+
+                int contentLength = Integer.parseInt(cl);
+                Debug.log("- contentLength: " + contentLength);
+                ConsoleProgressObserver progressObserver = new ConsoleProgressObserver(40);
+                progressObserver.setMax(contentLength);
+
+                File zf = new File(name + "/gtfs.zip");
+                FileOutputStream fos = new FileOutputStream(zf);
+
+                Util.downloadURLContent(gtfsURL, fos, progressObserver);
```

Comment:
You should check the modification time of the cached gtfs.zip against the modification time of the URL endpoint. If the latter isn't greater than the former, you don't need to redownload.

For extra points, you can create a version of `Util.downloadURLContent()` that takes an extra `pathToCachedVersion` arg, and then solve the general issue there instead of just the specific instance here.

Reply:
If the arg is `null` then don't check the modification date. Make an overloaded version of the method that has the same signature as before so you don't have to change calls to it everywhere.

Reply:
I've made this update. I thought through what it would take to have a Util function that serves this purpose - it's a great idea since this is very similar to Util.UpdateCacheIfNeeded. I ended up deciding not to do it, because the need to check gcloud in this instance, rather than the cache, leads to a bit too much complexity.